### DISCUSSION TITLE
M1 Chip support

### DISCUSSION THREAD

**Discussion by Pankaj003** [google/filament#5576] [Q&A]
Hi,
I've an application created using Qt and filament. 
OS:  macOS Monterey - 12.3.1
Filament: version 1.9.9 with backend as openGL(filament::Engine::Backend::OPENGL)
Build environment.: XCode - 13.4
I am making changes to support my application(dylib) on Mac M1 chip based systems.
Can someone please help me in understanding what changes are needed w.r.t filament library.
Some Questions:
1. Which version of filament works with M1 chip?
2. Will  backend as OPENGL work with M1 chip or we need to change it to METAL?
3. What other changes like project settings needed to support filament in my app, so that it can work on M1 chip based system?

Thanks

**Comment by romainguy** [comment]
You'll need a more recent version of Filament but open GL and metal will work on M1.

**Comment by Pankaj003** [comment]
Thanks for the update. Release binaries are built with x86_64 arch, how can I build with arch x86_64 and arm64 arch?

**Reply by bejado** [reply]
See the `-l` option in our `build.sh` script:
```
    -l
        Build arm64/x86_64 universal libraries.
        For iOS, this builds universal binaries for devices and the simulator (implies -s).
        For macOS, this builds universal binaries for both Apple silicon and Intel-based Macs.
```

**Comment by Pankaj003** [comment]
@bejado, @romainguy 
Hi,
I am able to build the universal libs(version: filament-1.22.2) on Mac using your suggestion. But after including new headers and libs I am getting lot of build errors in my project( earlier it used to be fine with version 1.9.9). I am not expert on Mac and looks like I am missing some setting required by filament. I've included c++17 in my project settings.
Could you please help me in resolving these errors, they are coming from different filament header files.
I am attaching error log file, error snapshot and my XCode project file, please have a look and let me know how I can resolve these build errors:

Thanks,

Error Snapshot:
![ErrorSnapShot](https://user-images.githubusercontent.com/8086656/171801582-c4cbc2fa-07a6-4af6-a241-1bfe4ad9ec7b.PNG)

Error Log file exported from XCode:
[Build target_2022-06-03T09-17-57.txt](https://github.com/google/filament/files/8830147/Build.target_2022-06-03T09-17-57.txt)

My XCode project Settings file:
[AI3D.xcodeproj.zip](https://github.com/google/filament/files/8830155/AI3D.xcodeproj.zip)

**Reply by romainguy** [reply]
Those errors are in the stdlib headers. Try a clean build, it’s often necessary after an Xcode or macOS update :/

**Comment by Pankaj003** [comment]
I did clean build and even physically deleted all the build folders, but still getting same errors...

**Comment by Pankaj003** [comment]
@bejado, @romainguy 
Hi,

I tried multiple things but nothing is working, getting same compilation errors. If I replace the filament version(v.1.22.2) with the older version of filament v.1.9.9(x86_64) then the project compiles well. But If I use the latest version filament1.22.2 build with universal(x86_64;arm64) then getting lots of compilation error as mentioned in the older comments. I am using Xcode13.4.1, I tried downgrading to Xcode12 but it is not supported on macOS Monterey.
Downloaded filament **x86_64 filament-v1.22.2-mac.tgz** too are not compiling in my project, throwing same build errors.
I've run out of ideas, could you please suggest how can I make filament compile correctly on the M1 chip machine?
May be I need to set some project settings either while building the filament or inside my Xcode project.
Please suggest....

Thanks,

**Reply by bejado** [reply]
Can you show how you're `#include`ing Filament?

**Comment by Pankaj003** [comment]
@bejado, @romainguy 
Thanks for checking this.

I am including filament header files in multiple .cpp and .h files. Below I am putting all of them. Currently I am using filament.v.1.9.1(x_86_64) and it is building fine. I am now trying to move to filament.v.1.23.0(universal build).
#include <filament/Texture.h>
#include <utils/Entity.h>
#include <filament/Engine.h>
#include <filament/Renderer.h>
#include <filament/IndexBuffer.h>
#include <filament/LightManager.h>
#include <filament/Material.h>
#include <filament/RenderableManager.h>
#include <filament/TransformManager.h>
#include <filament/Renderer.h>
#include <filament/Scene.h>
#include <filament/VertexBuffer.h>
#include <filament/View.h>
#include <filament/Viewport.h>
#include <filament/Fence.h>
#include <filament/SwapChain.h>
#include <filament/Camera.h>
#include <utils/EntityManager.h>
#include <utils/Path.h>
#include <filament/TextureSampler.h>
#include <filament/IndirectLight.h>
#include <filament/RenderTarget.h>
#include <filamat/MaterialBuilder.h>
#include <math/mat3.h>
#include <filament/Fence.h>

#include <filament/SwapChain.h>
#include <filament/Camera.h>
#include <utils/EntityManager.h>
#include <utils/Path.h>
#include <filament/TextureSampler.h>
#include <filament/IndirectLight.h>
#include <filament/RenderTarget.h>
#include <filamat/MaterialBuilder.h>
#include <filament/MaterialInstance.h>
#include <filament/VertexBuffer.h>
#include <filament/IndexBuffer.h>
#include <filament/Box.h>
#include <utils/Entity.h>

#include <filament/Scene.h>
#include <filament/RenderableManager.h>
#include <filament/TransformManager.h>
#include <utils/EntityManager.h>
#include <math/mat3.h>

#include <filament/Material.h>
#include <filament/Texture.h>
#include <filament/TextureSampler.h>

#include <filament/Texture.h>
#include <image/KtxBundle.h>
#include <image/KtxUtility.h>

#include <filament/Engine.h>
#include <filament/Material.h>
#include <filament/MaterialInstance.h>
#include <filament/Texture.h>
#include <filament/Skybox.h>

#include <image/KtxBundle.h>
#include <image/KtxUtility.h>

#include <utils/Path.h>
#include <filament/IndirectLight.h>

#include <filament/Engine.h>
#include <math/vec3.h>

#include <math/mat3.h>
#include <filament/Camera.h>

My Xcode project setting file:
[AI3D.xcodeproj.zip](https://github.com/google/filament/files/8849159/AI3D.xcodeproj.zip)

Error Log file exported from XCode:
[Build.target_2022-06-03T09-17-57.txt](https://github.com/google/filament/files/8849187/Build.target_2022-06-03T09-17-57.txt)

Thanks

**Comment by Pankaj003** [comment]
@bejado, @romainguy 
Hi,
One update, I built filament v.1.20.1 and it seems to be working fine, no compilation issues. Looks like somewhere between 1.20.1 and latest release, filament packaging have changed which may be causing these compilation errors.

Thanks

**Reply by bejado** [reply]
Thanks @Pankaj003, I'm not able to reproduce the compilation errors on my Intel-based Mac, but I'll try later with my M1 device.
